enhancements/network/persistent-ips-sdn-secondary-networks.md

```diff
@@ -0,0 +1,1020 @@
+- As an Openshift Virtualization user, I want to migrate (one of) my VMs to
+another node, preserving the networking configuration, so the IP-aware
+applications running in the VM are not impacted.
+- As an Openshift Virtualization user, I want to start Virtual Machines
+(currently stopped) with the original VM network configuration, so the IP-aware
+applications running in the VM are not impacted.
+- As an Openshift network administrator, I want to decide which networks can be
+subject to having persistent IP allocations surviving VM migration, and VM stop.
+
+### Goals
+
+- Manage Openshift Virtual Machines IP addresses from within the SDN solution
+provided by OVN-Kubernetes.
+- IPv4 range can be defined on OVN Kubernetes secondary network
+- VMs are able to get a IP from this subnet
+- The IP sticks to a VM during live-migration
+- The IP sticks to a VM even if it is temporarily shut down
+
+### Non-Goals
+
+- This enhancement's scope is limited to secondary networks of `layer2` or
+  `localnet` topologies; meaning any type of routed topology (e.g. the default
+  cluster network **or** routed topology secondary networks) IP allocation
+  should still follow the pod lifecycle.
+
+## Proposal
+
+### Workflow Description
+
+#### Creating a Virtual Machine
+The user initially requests a virtual machine (VM) to be created. If they
+desire their VM to have "sticky IPs" - i.e. the IP allocations to outlive the
+pod the VM runs on - the following requirements must be met:
+
+- the KubeVirt CR must have the required feature gate enabled: e.g.
+  `VMPersistentIPs`
+- the NAD holding the configuration for the secondary network must allow for
+persistent IPs - `{..., "allowPersistentIPs": true, ...}`
+
+If the KubeVirt feature gate is enabled, KubeVirt will create the `IPAMClaims`
+for each multus non-default network, thus declaring intent of having persistent
+IP allocations for those interfaces.
+This could be further improved by having KubeVirt only create those for multus
+non-default network having the `allowPersistentIPs` knob enabled - KubeVirt
+already reads `NetworkAttachmentDefinition`s. Refactoring it to read those via
+informers would be good, to alleviate the API load.
+
+The IPAM CNI plugin will then react accordingly and if the NAD configuration
+allows it, allocate IPs from its pool - which will be persisted in the
+`IPAMClaim` status. Finally, the CNI will configure the interface with these
+IP addresses.
+
+Please refer to the diagram below to better understand the proposed workflow
+for VM creation:
+1. the user provisions a VM object
+2. the KubeVirt controller creates an IPAMClaim for each multus non-default
+  network in the corresponding VMI spec
+3. OVN-K reacts to the creation of the `IPAMClaim`s. It will allocate IP
+  addresses from the required IP pools, and **only afterward** update the
+  corresponding `IPAMClaim` with the generated IPs. Users can only rely / use
+  the `IPAMClaim` status for informational purposes.
```

Comment:
I think it is worth mentioning here that OVN-K does not react to the creation of the `IPAMClaim`, it rather reacts to a pod request that features a `ipam-claim-reference` pointing to that IPAMClaim for one of the networks it is managing.

Reply:
Done.

Is this OK now @jcaamano ?